Grpc post processing throws OOM exception with large number of proto file

### Describe the bug

Importing a large set of proto files from a dependency like `io.envoyproxy.controlplane:api`  causes OOM.

```
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:generate-code (default) on project grpc: Quarkus code generation phase has failed: InvocationTargetException: Java heap space -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
```



### Expected behavior

Post-processing passed without tuning heap size

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

GrpcPostProcessing parsed all files in memory before post processing them. It should process files sequentially. I'll provide a MR to fix it.

--- a/extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcPostProcessing.java
+++ b/extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcPostProcessing.java
@@ -1,6 +1,13 @@
 package io.quarkus.grpc.deployment;
 
+import java.io.IOException;
+import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.function.Consumer;
 
 import org.eclipse.microprofile.config.Config;
 import org.jboss.logging.Logger;
@@ -43,6 +50,7 @@ public class GrpcPostProcessing {
 
     public void postprocess() {
         SourceRoot sr = new SourceRoot(root);
+        Map<Path, Path> changedFiles = new HashMap<Path, Path>();
         try {
             sr.parse("", new SourceRoot.Callback() {
                 @Override
@@ -54,9 +62,19 @@ public class GrpcPostProcessing {
                         if (unit.getPrimaryType().isPresent()) {
                             TypeDeclaration<?> type = unit.getPrimaryType().get();
                             postprocess(unit, type, context.config());
-                            return Result.SAVE;
-                        }
 
+                            // save to a tempory file first, then move all temporary unit files at the same time
+                            try {
+                                unit.setStorage(Files.createTempFile(null, null),
+                                        sr.getParserConfiguration().getCharacterEncoding())
+                                        .getStorage().get().save(sr.getPrinter());
+                            } catch (IOException ex) {
+                                throw new RuntimeException(ex);
+                            }
+
+                            changedFiles.put(unit.getStorage().get().getPath(), absolutePath);
+                            return Result.DONT_SAVE;
+                        }
                     } else {
                         // Compilation issue - report and skip
                         log.errorf(
@@ -68,9 +86,32 @@ public class GrpcPostProcessing {
                     return Result.DONT_SAVE;
                 }
             });
+
+            changedFiles.entrySet().stream().forEach(new Consumer<Entry<Path, Path>>() {
+                @Override
+                public void accept(Entry<Path, Path> entry) {
+                    try {
+                        Files.move(entry.getKey(), entry.getValue(), StandardCopyOption.REPLACE_EXISTING);
+                    } catch (IOException ex) {
+                        throw new RuntimeException(ex);
+                    }
+                }
+            });
+            changedFiles.clear();
+
         } catch (Exception e) {
             // read issue, report and exit
             log.error("Unable to parse the classes generated using protoc - skipping gRPC post processing", e);
+        } finally {
+            changedFiles.entrySet().stream().forEach(new Consumer<Entry<Path, Path>>() {
+                @Override
+                public void accept(Entry<Path, Path> e) {
+                    try {
+                        Files.deleteIfExists(e.getKey());
+                    } catch (IOException discard) {
+                    }
+                }
+            });
         }
     }
 

--- a/extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcPostProcessing.java
+++ b/extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcPostProcessing.java
@@ -1,14 +1,10 @@
 package io.quarkus.grpc.deployment;
 
-import java.nio.charset.StandardCharsets;
-import java.nio.file.Files;
 import java.nio.file.Path;
-import java.util.List;
 
 import org.eclipse.microprofile.config.Config;
 import org.jboss.logging.Logger;
 
-import com.github.javaparser.ParseResult;
 import com.github.javaparser.ast.CompilationUnit;
 import com.github.javaparser.ast.Modifier;
 import com.github.javaparser.ast.body.ClassOrInterfaceDeclaration;
@@ -48,26 +44,30 @@ public class GrpcPostProcessing {
     public void postprocess() {
         SourceRoot sr = new SourceRoot(root);
         try {
-            // Parse all files from root
-            List<ParseResult<CompilationUnit>> results = sr.tryToParse();
-            for (ParseResult<CompilationUnit> result : results) {
-                if (result.isSuccessful()) {
-                    CompilationUnit unit = result.getResult().orElseThrow(); // the parsing succeed, so we can retrieve the cu
-                    CompilationUnit.Storage storage = unit.getStorage().orElseThrow(); // we read from the FS, so we have a storage
-                    if (unit.getPrimaryType().isPresent()) {
-                        TypeDeclaration<?> type = unit.getPrimaryType().get();
-                        postprocess(unit, type, context.config());
-                    }
-                    Files.write(storage.getPath(), List.of(unit.toString()), StandardCharsets.UTF_8);
-                } else {
-                    // Compilation issue - report and skip
-                    log.errorf(
-                            "Unable to parse a class generated using protoc, skipping post-processing for this " +
-                                    "file. Reported problems are %s",
-                            result.toString());
-                }
+            sr.parse("", new SourceRoot.Callback() {
+                @Override
+                public com.github.javaparser.utils.SourceRoot.Callback.Result process(Path localPath, Path absolutePath,
+                        com.github.javaparser.ParseResult<CompilationUnit> result) {
+                    if (result.isSuccessful()) {
+                        CompilationUnit unit = result.getResult().orElseThrow(); // the parsing succeed, so we can retrieve the cu
 
-            }
+                        if (unit.getPrimaryType().isPresent()) {
+                            TypeDeclaration<?> type = unit.getPrimaryType().get();
+                            postprocess(unit, type, context.config());
+                            return Result.SAVE;
+                        }
+
+                    } else {
+                        // Compilation issue - report and skip
+                        log.errorf(
+                                "Unable to parse a class generated using protoc, skipping post-processing for this " +
+                                        "file. Reported problems are %s",
+                                result.toString());
+                    }
+
+                    return Result.DONT_SAVE;
+                }
+            });
         } catch (Exception e) {
             // read issue, report and exit
             log.error("Unable to parse the classes generated using protoc - skipping gRPC post processing", e);

